adding more_like_this query dsl docs (#9746)
opensearch-project/documentation-website: _query-dsl/specialized/more-like-this.md

{
  "query": {
    "more_like_this": {
      "fields": ["content"],
      "like": "sand dunes hidden oases",
      "min_term_freq": 1,
      "min_doc_freq": 1
    }
  }
}
```
{% include copy-curl.html %}

The terms that get selected by `more_like_this` are: `sand`, `dunes`, `hidden`, and `oases`, which match one document:

```json
{
  ...
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 2.5893893,
    "hits": [
      {
        "_index": "articles-optimized",
        "_id": "1",
        "_score": 2.5893893,
        "_source": {
          "title": "Sahara Desert Expeditions",
          "content": "An ocean of golden sand and hidden oases."
        }
      }
    ]
  }
}
```

## Example using multiple documents and text input in `like`

The `more_like_this` query allows you to provide multiple sources in the `like` parameter. You can combine free text with documents from the index. This is useful if you want the search to combine relevance signals from several examples.

In the following example:
- A new artificial document is provided directly.
- An existing document with ID `5` from the `heroes` index is included.

```json

> GET /_search

Reviewer: Can we add a sample response for this request as well? It helps

          "quote": "Even in the darkest times, hope cuts through."
        }
      }
    ]
  }
}
```

Use this pattern when you want to boost results based on a new concept that is not yet fully indexed, but also want to combine it with knowledge from existing indexed documents.
{: .note}

# Parameters

The only required parameter for a `more_like_this` query is `like`. The rest of the parameters have default values but allow fine-tuning. Parameters fall into three main categories:

## Document Input Parameters

| Parameter | Required/Optional | Data Type | Description |
| `like`| Required| Array of strings or objects | Defines the text or documents to find similar documents for. You can input free text, real documents from the index, or even artificial documents. The analyzer associated with the field will process the text unless you override it. |
| `unlike`| Optional| Array of strings or objects | Provides text or documents whose terms should be *excluded* from influencing the query. Useful for specifying negative examples.|
| `fields`| Optional| Array of strings| Lists fields to use when analyzing text. If not specified, all fields are used. |

## Term Selection Parameters

> | Parameter | Required/Optional | Data Type| Description|

Reviewer 2: ```suggestion
| Parameter | Required/Optional | Data Type| Description|
| :--- |  :--- |  :--- |  :--- | 
```

| `max_query_terms` | Optional| Integer| Sets the maximum number of terms to select from the input. A higher value increases precision but slows down execution. Default is `25`. |
| `min_term_freq` | Optional| Integer| Terms appearing fewer times than this in the input will be ignored. Default is `2`.|
| `min_doc_freq`| Optional| Integer| Terms that appear in fewer documents than this value will be ignored. Default is `5`.|

> | `max_doc_freq`| Optional| Integer| Terms appearing in more documents than this limit are ignored. Useful for avoiding very common words. Default is unlimited (`2147483647`). |

Reviewer 2: Does this number correspond to a power of 2 minus something? Might be easier to list it in this form.

Author: @kolchfa-aws  It can be written as 2^31 - 1, do you think this would be better?

Reviewer 2: I think it's more readable. `2<sup>31</sup> - 1`

Author: Thats updated now

| `min_word_length` | Optional| Integer| Ignore words shorter than this value. Default is `0`.|
| `max_word_length` | Optional| Integer| Ignore words longer than this value. Default is unlimited. |
| `stop_words`| Optional| Array of strings | Defines a list of words that are ignored completely when selecting terms.|
| `analyzer`| Optional| String | The custom analyzer to use for processing input text. Defaults to the analyzer of the first field listed in `fields`.|

> ## Query Formation Parameters

Reviewer 2: ```suggestion
## Query formation parameters
```

> | Parameter | Required/Optional | Data Type| Description|

Reviewer 2: ```suggestion
| Parameter | Required/Optional | Data Type| Description|
| :--- |  :--- |  :--- |  :--- | 
```

> | `max_query_terms` | Optional| Integer| Sets the maximum number of terms to select from the input. A higher value increases precision but slows down execution. Default is `25`. |

Reviewer 2: This is a duplicate list from the previous table. It should contain query formation parameters `minimum_should_match`, `fail_on_unsupported_field`, `boost_terms`, `include`, and `boost`.

Author: Absolutely, was there as a place holder and forgot to update it, thats done now

---
layout: default
title: More like this
parent: Specialized queries
nav_order: 45
has_math: false
---

# More like this

Use a `more_like_this` query to find documents that are similar to one or more given documents. This is useful for recommendation engines, content discovery, and identifying related items in a dataset.

The `more_like_this` query analyzes the input documents or texts and selects terms that best characterize them. It then searches for other documents that contain those significant terms.

## Prerequisites

Before you use a `more_like_this` query, ensure that the fields you target are indexed and their data type is either [`text`]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/text/) or [`keyword`]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/keyword/). 

If you reference documents in the `like` section, OpenSearch needs access to their content. This is typically done through the `_source` field, which is enabled by default. If `_source` is disabled, you must either store the fields individually or configure them to save [`term_vector`]({{site.url}}{{site.baseurl}}/field-types/mapping-parameters/term-vector/) data. 

> Saving [`term_vector`]({{site.url}}{{site.baseurl}}/field-types/mapping-parameters/term-vector/) information when indexing documents can greatly accelerate `more_like_this` queries, because the engine can directly retrieve the important terms without re-analyzing the field text at query time.

Reviewer 3: ```suggestion
Saving [`term_vector`]({{site.url}}{{site.baseurl}}/field-types/mapping-parameters/term-vector/) information when indexing documents can greatly accelerate `more_like_this` queries because the engine can directly retrieve the important terms without reanalyzing the field text at query time.
```

    "hits": [
      {
        "_index": "articles-basic",
        "_id": "2",
        "_score": 1.9616582,
        "_source": {
          "title": "Amazon Rainforest Tour",
          "content": "Dense jungle and exotic wildlife."
        }
      }
    ]
  }
}
```

## Example: Term vector optimization

Create an index named `articles-optimized` using the following mapping:

```json
PUT /articles-optimized
{
  "mappings": {
    "properties": {
      "title": {
        "type": "text",
        "term_vector": "with_positions_offsets"
      },
      "content": {
        "type": "text",
        "term_vector": "with_positions_offsets"
      }
    }
  }
}
```
{% include copy-curl.html %}

Insert sample documents into the optimized index:

```json
POST /articles-optimized/_bulk
{ "index": { "_id": "a1" } }
{ "name": "Diana", "alias": "Wonder Woman", "quote": "Justice will come when it is deserved." }
{ "index": { "_id": "a2" } }
{ "name": "Clark", "alias": "Superman", "quote": "Even in the darkest times, hope cuts through." }
{ "index": { "_id": "a3" } }
{ "name": "Bruce", "alias": "Batman", "quote": "I am vengeance. I am the night. I am Batman!" }
```
{% include copy-curl.html %}

> Find documents similar to `dark night` using the following request:

Reviewer 3: Is this accurately stated, or do we mean something like "Find documents similar to those containing `dark night`..." or "Find documents similar to those containing `dark` and `night`..."?

> Find documents similar to `dark` and `night` using the following request:

Reviewer 3: @AntonEliatra The documents themselves can't literally be similar to `dark` and `night`. Do we mean "Find documents similar to those containing `dark` and `night`..."?

Author: @natebower I've updated this now to:

Find documents where the `quote` field contains terms similar to "dark" and "night"

> Find documents where the `quote` field contains terms similar to "dark" and "night":

Reviewer 3: ```suggestion
Find documents in which the `quote` field contains terms similar to "dark" and "night":
```

      "relation": "eq"
    },
    "max_score": 2.140194,
    "hits": [
      {
        "_index": "articles-optimized",
        "_id": "a1",
        "_score": 2.140194,
        "_source": {
          "name": "Diana",
          "alias": "Wonder Woman",
          "quote": "Justice will come when it is deserved."
        }
      },
      {
        "_index": "articles-optimized",
        "_id": "a2",
        "_score": 1.1596459,
        "_source": {
          "name": "Clark",
          "alias": "Superman",
          "quote": "Even in the darkest times, hope cuts through."
        }
      }
    ]
  }
}
```

Use this pattern when you want to boost results based on a new concept that is not yet fully indexed, but also want to combine it with knowledge from existing indexed documents.
{: .note}

# Parameters

The only required parameter for a `more_like_this` query is `like`. The rest of the parameters have default values but allow fine-tuning. Parameters fall into the following main categories.

## Document input parameters

The following table specifies document input parameters.

| Parameter | Required/Optional | Data type | Description |
| :--- |  :--- |  :--- |  :--- | 

> | `like`| Required| Array of strings or objects | Defines the text or documents to find similar documents for. You can input free text, real documents from the index, or artificial documents. The analyzer associated with the field processes the text unless overridden. |

Reviewer 3: ```suggestion
| `like`| Required| Array of strings or objects | Defines the text or documents for which to find similar documents. You can input free text, real documents from the index, or artificial documents. The analyzer associated with the field processes the text unless overridden. |
```
